### DISCUSSION TITLE
RFC:Using OpenDAL to make application metadata persistent

### DISCUSSION THREAD

**Discussion by bokket** [SeaQL/sea-orm#1557] [General]
OpenDAL is an abstract data access layer written by rust,the goal is:acessing data freely, painlessly, and efficiently.
More than a dozen storage services are already supported and used in production as the data layer of [databend](https://github.com/datafuselabs/databend),[mozilla/sccache](https://github.com/mozilla/sccache/),[GreptimeDB](https://github.com/GreptimeTeam/greptimedb)....
Ref to idea:[A media-centric CMS based on SeaORM & Seaography](https://github.com/SeaQL/summer-of-code/tree/main/2023#4-a-media-centric-cms-based-on-seaorm--seaography).
Here are a few steps to implement:
1. No need admin panel, we design a reasonable directory format and glorify it. All metadata will be stored in the `toml file` format, so it might actually look like a `Cargo file`.
2. When we upload images and short videos from the database or locally, it will store them in opendel.This can be horizontally extended to different applications such as gdive, onedive,Dropbox and so on.This is no longer limited to admin who need to access the server via ssh in order to uploadBut these implementations are still in the `working on` stage of opendal's [roadmap](https://github.com/apache/incubator-opendal/issues/5).
3. [Listen to file system events and update the database index immediately](https://github.com/emcrisostomo/fswatch) .Not sure is opendal upstream may want to implement the function, should need to be [redone]( https://github.com/ascclemens/fswatch-sys ) in this idea.
4. Need a database to store the metadata.Maybe?OpenDal full support to be [the data layer of the database](https://github.com/apache/incubator-opendal#opendal---).But this is a lot of work.Using SQLite as a datastore is not required.The external dependency is in opendal, which updates the index of metadata in real time.
5. Deliver a minimal UI.

**Comment by bokket** [comment]
cc @billy1624

**Comment by billy1624** [comment]
Hey @bokket, sorry for coming in late. I think this is out of the scope of SeaORM. But yes, OpenDAL could be used in [A media-centric CMS based on SeaORM & Seaography](https://github.com/SeaQL/summer-of-code/tree/main/2023#4-a-media-centric-cms-based-on-seaorm--seaography).
